_posts/2022-05-19-kotlin-context-receivers.markdown

```diff
@@ -0,0 +1,504 @@
+> reference to the context receiver.
+
+### Use Case: Composable DSLs
+
+In the following example, we are using the `buildJsonObject` function, which is provided by the kotlinx.serialization
+library. Its usage is not the most convenient:
+
+```kotlin
+val jsonObject = buildJsonObject {
+  put("clubName", "Kotlin User Group")
+  putJsonArray("members") {
+    add(buildJsonObject {
+      put("name", "Bob")
+      put("age", 12)
+    })
+    add(buildJsonObject {
+      put("name", "Sue")
+      put("age", 34)
+    })
+  }
+}
+```
+
+We would like to create a DSL that will simplify above example and delegate to the `buildJsonObject` function. Prior
+to context receivers this was possible only by using member extensions: create a custom class, that would either
+duplicate all members of the `JsonObjectBuilder` class or subclass it (if we are lucky, and it is open). Either way,
+lots of code ceremony is involved and the result is still very inflexible.
+
+Thanks to context receivers, this is now very easy to do with top level functions:
+
+```kotlin
+fun json(action: JsonObjectBuilder.() -> Unit) = buildJsonObject(action)
+
+context(JsonObjectBuilder)
+infix fun String.by(value: String) = put(this, value)
+
+context(JsonObjectBuilder)
+infix fun String.by(value: Number) = put(this, value)
+
+context(JsonObjectBuilder)
+infix fun String.by(values: List<JsonElement>) = putJsonArray(this) {
+  for (value in values) add(value)
+}
+
+
+val json2 = json {
+  "clubName" by "Kotlin User Group"
+  "members" by listOf(
+    json {
+      "name" by "Bob"
+      "age" by 12
+    },
+    json {
+      "name" by "Sue"
+      "age" by 34
+    }
+  )
+}
+```
+
+### Use Case: Dependency Injection
+
+Consider the following simple `SearchController`. It is supposed to be managed by a dependency injection framework,
+that will inject 2 dependencies for us, a `Urls` object and an `I18n` object.
+
+This SearchController is rendering a very simple HTML page, and for doing so, it is delegating parts of its
+implementation to a global top-level `homebreadcrumb` function.
+
+```kotlin
+interface Urls {
+  val homeUrl: String
+  val searchUrl: String
+}
+
+interface I18n {
+  val homeTitle: String
+  val searchTitle: String
+}
+
+class SearchController(
+  private val urls: Urls,
+  private val i18n: I18n,
+) {
+
+  fun render(): String = """
+    <html>
+      <head>
+        <title>${i18n.searchTitle}</title>
+      </head>
+      <body>
+        ${homeBreadcrumb(urls, i18n)}
+        <h1>${i18n.searchTitle}</h1>
+      </body>
+    </html>
+  """.trimIndent()
+}
+
+fun homeBreadcrumb(urls: Urls, i18n: I18n): String = """<a href="${urls.homeUrl}">${i18n.homeTitle}</a>"""
+```
+
+The problem in above example is that dependency injection does not play well with top level functions. The `urls` and
+`i18n` dependencies cannot be injected automatically for us, instead we need to provide them explicitly as arguments.
```

Comment:
change to 

The `urls` and
`i18n` dependencies cannot be injected automatically into the homebreadcrumb function for us, instead we need to provide them explicitly as arguments